File: crates/backend/web/runtime/js/virtualizer.js
// Virtualizer — JS-side scroll handler + visible-range diff for the
// web backend's `Primitive::Virtualizer`. Owns the scroll listener
// so the wasm-bindgen boundary is crossed only when items
// enter/leave the visible window, not on every scroll tick.
//
// The framework's `backend-web::create_virtualizer` injects this
// file's contents into a `<script>` tag on first use, then constructs
// an instance and hands it the Rust callbacks via wasm-bindgen-
// wrapped closures.
//
// # Lanes / grid
//
// The engine is lane-based. A *list* is one lane: each item owns a
// full cross-axis line. A *grid* is N lanes: item `i` lands in lane
// `i % L` of grid-row `i / L`. Range math runs over grid-rows (a
// prefix-sum of per-row main-axis extents), then expands the visible
// grid-rows to the item indices `[rowStart*L, (rowEnd+1)*L)`. The
// list path is exactly the grid path with `L === 1`, so there is no
// separate code path and no list-vs-grid branch in the hot loop.
//
// `lanesFixed` (a column count, 1 = list) and `lanesMinCross` (a
// responsive minimum lane size, CSS `auto-fill`-style) are mutually
// exclusive; the Rust side sets exactly one. `mainSpacing` is the gap
// between grid-rows along the scroll axis, `crossSpacing` the gap
// between lanes.
//
// Type sketch (JSDoc — keeps editor type-aware without a TS build):
//
// /**
//  * @typedef {Object} VirtualizerCallbacks
//  * @property {() => number} itemCount
//  * @property {(idx: number) => number} itemKey
//  * @property {(idx: number) => number} itemSize  Main-axis extent.
//  * @property {(idx: number) => [Element, number]} mountItem
//  *   Returns [DOM node, scopeId]. Scope id is opaque to JS.
//  * @property {(scopeId: number) => void} releaseItem
//  * @property {(scopeId: number, size: number) => void} setMeasuredSize
//  *   Backend->framework: notify a measured size change.
//  * @property {boolean} measureSizes
//  * @property {number} overscan
//  * @property {boolean} horizontal
//  * @property {number} [lanesFixed]   Fixed lane count (>=1).
//  * @property {number} [lanesMinCross] Responsive min lane cross size.
//  * @property {number} [mainSpacing]
//  * @property {number} [crossSpacing]
//  */

(function () {
    if (typeof window === 'undefined') return;
    if (window.__idealystVirtualizer) return; // idempotent inject

    class Virtualizer {
        /**
         * @param {HTMLElement} container — the outer scroller
         * @param {Object} cb — callbacks bundle (see typedef above)
         */
        constructor(container, cb) {
            this.container = container;
            this.cb = cb;
            this.horizontal = !!cb.horizontal;
            this.mainSpacing = +cb.mainSpacing || 0;
            this.crossSpacing = +cb.crossSpacing || 0;
            // Lane config: exactly one of the two is meaningful.
            this.lanesFixed = cb.lanesFixed != null ? Math.max(1, cb.lanesFixed | 0) : null;
            this.lanesMinCross = cb.lanesMinCross != null ? +cb.lanesMinCross : null;
            // Resolved at layout time from the container cross extent.
            this.lanes = this.lanesFixed || 1;

            // Outer scroller (the container) holds an inner spacer
            // sized to the total content extent so the scrollbar is
            // accurate. Mounted items absolute-positioned within.
            container.style.overflow = this.horizontal ? 'auto hidden' : 'hidden auto';
            container.style.position = container.style.position || 'relative';

            this.spacer = document.createElement('div');
            this.spacer.style.position = 'relative';
            this.spacer.style.width = this.horizontal ? '0px' : '100%';
            this.spacer.style.height = this.horizontal ? '100%' : '0px';
            // The container might be a flex parent (the framework's
            // default class sets `display: flex; flex-direction: column`
            // on every styled view). Flex children won't honor an
            // explicit main-axis size if `flex-shrink` is the default
            // 1 — the spacer would collapse to fit the viewport
            // instead of forcing a scrollable extent. Pin shrink to 0
            // and basis to auto so the spacer's explicit width/height
            // wins.
            this.spacer.style.flexShrink = '0';
            this.spacer.style.flexBasis = 'auto';
            container.appendChild(this.spacer);

            /** @type {Map<number, {node: Element, scopeId: number, idx: number, key: number, size: number, observer?: ResizeObserver}>} */
            this.mountedByIdx = new Map();
            /** @type {Map<number, number>} Map key -> mountedByIdx idx, for keyed-diff preservation. */
            this.keyToIdx = new Map();

            this.lastStart = -1;
            this.lastEnd = -1;
            this.totalSize = 0;
            // prefixRow[r] = main-axis offset of grid-row r (cumulative
            // row extents + inter-row spacing). Length = rowCount + 1.
            this.prefixRow = [0];
            this.rowCount = 0;
            // Resolved cross-axis extent per lane, set in refresh().
            this.laneCross = 0;

            // Store the scroll handler as a named property so
            // `release()` can detach it. An anonymous arrow would
            // be impossible to removeEventListener.
            this._scrollHandler = () => {
                if (this._released) return;
                this.update();
            };
            container.addEventListener('scroll', this._scrollHandler, { passive: true });
            // Also re-update on container resize so viewport changes
            // trigger a re-window. A resize can also change the
            // resolved lane count (AutoFit) or lane width — if so we
            // need a full `refresh()`, not just a range `update()`.
            if (typeof ResizeObserver !== 'undefined') {
                this._containerObserver = new ResizeObserver(() => {
                    if (this._released) return;
                    const prevLanes = this.lanes;
                    const prevCross = this.laneCross;
                    if (this._resolveLanes() !== prevLanes) {
                        this.refresh();
                    } else if (Math.abs(this.laneCross - prevCross) > 0.5) {
                        // Lane count steady but each lane's width moved
                        // (container resized): reposition without a full
                        // remount by re-running the range pass.
                        this.update();
                    } else {
                        this.update();
                    }
                });
                this._containerObserver.observe(container);
            }
            this._released = false;

            // Defer the initial mount pass to a microtask: the Rust
            // side that constructed us still holds a `borrow_mut` on
            // the WebBackend RefCell. If we mount synchronously here,
            // `mountItem()` would re-enter the same RefCell from a
            // different call chain and trigger a "RefCell already
            // borrowed" panic.
            queueMicrotask(() => {
                if (this._released) return;
                this.refresh();
            });
        }

        /** The container's cross-axis extent (perpendicular to scroll). */
        _crossExtent() {
            return this.horizontal ? this.container.clientHeight : this.container.clientWidth;
        }

        /**
         * Resolve the concrete lane count + per-lane cross extent from
         * the container's current cross size. Returns the lane count
         * (also stored on `this.lanes`); updates `this.laneCross`.
         */
        _resolveLanes() {
            const cross = this._crossExtent();
            let lanes;
            if (this.lanesFixed != null) {
                lanes = this.lanesFixed;
            } else if (this.lanesMinCross != null && this.lanesMinCross > 0 && cross > 0) {
                // Largest N with N*min + (N-1)*gap <= cross.
                lanes = Math.floor((cross + this.crossSpacing) / (this.lanesMinCross + this.crossSpacing));
                lanes = Math.max(1, lanes);
            } else {
                lanes = 1;
            }
            this.lanes = lanes;
            // Per-lane cross extent: divide the remaining space after
            // inter-lane gaps. Guard a zero/unknown container.
            this.laneCross = lanes > 0 && cross > 0
                ? (cross - (lanes - 1) * this.crossSpacing) / lanes
                : 0;
            return lanes;
        }

        /** Main-axis extent of grid-row `r` = max item size across its lanes. */
        _rowExtent(r, n) {
            const L = this.lanes;
            let ext = 0;
            for (let lane = 0; lane < L; lane++) {
                const idx = r * L + lane;
                if (idx >= n) break;
                const s = this.cb.itemSize(idx);
                if (s > ext) ext = s;
            }
            return ext;
        }

        /** Recompute lanes + prefix sums + spacer extent + visible range from scratch. */
        refresh() {
            if (this._released) return;
            this._resolveLanes();
            const L = this.lanes;
            const n = this.cb.itemCount();
            this.rowCount = L > 0 ? Math.ceil(n / L) : 0;

            // Prefix-sum over grid-rows. prefixRow[r+1] = prefixRow[r]
            // + rowExtent(r) + mainSpacing. Trailing spacing is trimmed
            // from the total content extent below.
            this.prefixRow = new Array(this.rowCount + 1);
            this.prefixRow[0] = 0;
            for (let r = 0; r < this.rowCount; r++) {
                this.prefixRow[r + 1] = this.prefixRow[r] + this._rowExtent(r, n) + this.mainSpacing;
            }
            this.totalSize = this.rowCount > 0
                ? this.prefixRow[this.rowCount] - this.mainSpacing
                : 0;
            if (this.horizontal) {
                this.spacer.style.width = this.totalSize + 'px';
            } else {
                this.spacer.style.height = this.totalSize + 'px';
            }

            // Rebuild keyToIdx from current data.
            this.keyToIdx.clear();
            for (let i = 0; i < n; i++) {
                this.keyToIdx.set(this.cb.itemKey(i), i);
            }

            // Reconcile mounted set against the new data in two
            // passes. Phase 1: unmount any entry whose key is gone
            // (the underlying item was removed) — this is the only
            // place that drops DOM + per-item Rust scopes.
            for (const [oldIdx, entry] of this.mountedByIdx) {
                if (!this.keyToIdx.has(entry.key)) {
                    this._unmountEntry(oldIdx);
                }
            }
            // Phase 2: re-key every survivor by its NEW idx. Two
            // groups:
            //   - newIdx === entry.idx → item stayed in place; keep
            //     the existing map entry untouched.
            //   - newIdx !== entry.idx → item moved (reorder); we
            //     need to move it to a new map slot AND update its
            //     `entry.idx` field so subsequent `update()` calls
            //     position it correctly.
            //
            // Snapshot first so we can iterate while mutating the
            // map. `entries()` returns a live iterator on Map, but
            // we'd be deleting + inserting under the same iteration.
            const snapshot = Array.from(this.mountedByIdx.entries());
            const moved = [];
            for (const [oldIdx, entry] of snapshot) {
                const newIdx = this.keyToIdx.get(entry.key);
                if (newIdx === undefined) {
                    // Already unmounted in phase 1. Nothing to do.
                    continue;
                }
                if (newIdx !== oldIdx) {
                    this.mountedByIdx.delete(oldIdx);
                    moved.push({ entry, newIdx });
                }
            }
            for (const { entry, newIdx } of moved) {
                entry.idx = newIdx;
                this.mountedByIdx.set(newIdx, entry);
            }
            this.lastStart = -1;
            this.lastEnd = -1;
            this.update();
        }

        /**
         * Recompute visible range and apply mount/unmount diff.
         */
        update() {
            if (this._released) return;
            const L = this.lanes;
            const scroll = this.horizontal ? this.container.scrollLeft : this.container.scrollTop;
            const viewport = this.horizontal ? this.container.clientWidth : this.container.clientHeight;
            const buffer = viewport * (this.cb.overscan || 1.0);

            const startOffset = scroll - buffer;
            const endOffset = scroll + viewport + buffer;

            // Binary-search prefix-row sums for the first/last grid-row
            // overlapping the buffered window, then expand to item
            // indices. A grid-row of L lanes covers L consecutive
            // item indices.
            const n = this.cb.itemCount();
            const rowStart = this._findRowAtOffset(Math.max(0, startOffset));
            const rowEnd = Math.min(
                this.rowCount - 1,
                this._findRowAtOffset(Math.max(0, endOffset))
            );
            const start = rowStart * L;
            const end = Math.min(n - 1, (rowEnd + 1) * L - 1);

            if (start === this.lastStart && end === this.lastEnd) {
                // Range steady, but lane geometry may have shifted on a
                // resize — reposition mounted entries cheaply.
                for (const [idx, entry] of this.mountedByIdx) {
                    this._positionEntry(idx, entry);
                }
                return;
            }
            this.lastStart = start;
            this.lastEnd = end;

            // Unmount items outside [start, end].
            for (const [idx, _] of this.mountedByIdx) {
                if (idx < start || idx > end) {
                    this._unmountEntry(idx);
                }
            }

            // Mount items inside [start, end] not already mounted.
            for (let idx = start; idx <= end; idx++) {
                if (this.mountedByIdx.has(idx)) continue;
                this._mountIndex(idx);
            }

            // Reposition all currently-mounted entries (cheap; they
            // already exist, we just set top/left + cross size).
            for (const [idx, entry] of this.mountedByIdx) {
                this._positionEntry(idx, entry);
            }
        }

        _mountIndex(idx) {
            const [node, scopeId] = this.cb.mountItem(idx);
            const key = this.cb.itemKey(idx);
            const size = this.cb.itemSize(idx);
            node.style.position = 'absolute';
            this.spacer.appendChild(node);
            const entry = { node, scopeId, idx, key, size };
            this.mountedByIdx.set(idx, entry);
            this.keyToIdx.set(key, idx);
            this._positionEntry(idx, entry);

            // If we measure sizes, install a ResizeObserver. On
            // rendered-size change, push the new value back to Rust
            // and ourselves, then refresh layout. We observe the
            // main-axis dimension only — the cross axis is pinned to
            // the lane width.
            if (this.cb.measureSizes && typeof ResizeObserver !== 'undefined') {
                const obs = new ResizeObserver(() => {
                    const rect = node.getBoundingClientRect();
                    const newSize = this.horizontal ? rect.width : rect.height;
                    if (Math.abs(newSize - entry.size) < 0.5) return;
                    entry.size = newSize;
                    this.cb.setMeasuredSize(scopeId, newSize);
                    this.refresh();
                });
                obs.observe(node);
                entry.observer = obs;
            }
        }

        _unmountEntry(idx) {
            const entry = this.mountedByIdx.get(idx);
            if (!entry) return;
            if (entry.observer) entry.observer.disconnect();
            if (entry.node.parentNode === this.spacer) {
                this.spacer.removeChild(entry.node);
            }
            this.cb.releaseItem(entry.scopeId);
            this.mountedByIdx.delete(idx);
        }

        /**
         * Place an entry at its (main, cross) position. Main offset
         * comes from its grid-row's prefix sum; cross offset + size
         * from its lane. In single-lane (list) mode the item fills the
         * cross axis exactly as before.
         */
        _positionEntry(idx, entry) {
            const L = this.lanes;
            const row = (idx / L) | 0;
            const lane = idx - row * L;
            const main = this.prefixRow[row] || 0;
            const crossOff = lane * (this.laneCross + this.crossSpacing);
            const node = entry.node;
            if (this.horizontal) {
                node.style.left = main + 'px';
                if (L > 1) {
                    node.style.top = crossOff + 'px';
                    node.style.height = this.laneCross + 'px';
                } else {
                    node.style.top = '0';
                    node.style.height = '100%';
                }
            } else {
                node.style.top = main + 'px';
                if (L > 1) {
                    node.style.left = crossOff + 'px';
                    node.style.width = this.laneCross + 'px';
                } else {
                    node.style.left = '0';
                    node.style.width = '100%';
                }
            }
        }

        /**
         * Find the first grid-row `r` such that prefixRow[r+1] > offset.
         * Binary search. Clamps to [0, rowCount-1].
         */
        _findRowAtOffset(offset) {
            let lo = 0;
            let hi = this.prefixRow.length - 1;
            while (lo < hi) {
                const mid = (lo + hi) >> 1;
                if (this.prefixRow[mid + 1] > offset) {
                    hi = mid;
                } else {
                    lo = mid + 1;
                }
            }
            return lo;
        }

        /** Called by Rust when data changes (item_count effect fires).
         * Deferred to a microtask so we don't re-enter the
         * `WebBackend` RefCell while the framework still holds
         * `borrow_mut` from the call site that fired the data-change
         * effect. */
        dataChanged() {
            if (this._released) return;
            queueMicrotask(() => {
                if (this._released) return;
                this.refresh();
            });
        }

        /** Called by Rust from `release_virtualizer` when the
         * surrounding scope (a `when` branch, a `switch` arm, the
         * containing `Owner`) drops. Detaches every DOM listener,
         * disconnects all observers, unmounts everything currently
         * mounted, and flips `_released` so any late-firing handlers
         * (queued scroll/resize events, layout-change callbacks)
         * short-circuit instead of calling back into Rust closures
         * that may already have had their captured Signal scope
         * dropped.
         *
         * After `release()` the instance is inert; callers should
         * drop their references so JS can GC the instance + its
         * attached `_rust_cb_*` Closure wrappers. */
        release() {
            // Idempotent: a separate `_releasedFully` flag tracks
            // whether the full teardown has run. The Rust side may
            // set `_released = true` synchronously before invoking
            // `release()` from a microtask (see `release` in the
            // Rust virtualizer module). The `_released` flag is the
            // event-guard for queued listeners; this flag prevents
            // double-running the unmount loop.
            if (this._releasedFully) return;
            this._releasedFully = true;
            this._released = true;
            // 1. Stop listening for new scroll/resize events.
            this.container.removeEventListener('scroll', this._scrollHandler);
            if (this._containerObserver) {
                this._containerObserver.disconnect();
                this._containerObserver = null;
            }
            // 2. Unmount everything currently in the window. This
            //    detaches each item's ResizeObserver too.
            //    `_unmountEntry` calls `releaseItem(scopeId)` back
            //    into Rust to drop the per-item Scope. By this
            //    point the framework's outer `borrow_mut()` has
            //    been released (the Rust caller microtask-defers
            //    this release call), so per-item Scope drops are
            //    free to re-borrow the backend via
            //    `on_node_unstyled` / similar paths.
            for (const idx of Array.from(this.mountedByIdx.keys())) {
                this._unmountEntry(idx);
            }
            // 3. Clear the spacer so the DOM tree is empty too.
            if (this.spacer && this.spacer.parentNode === this.container) {
                this.container.removeChild(this.spacer);
            }
        }
    }

    window.__idealystVirtualizer = Virtualizer;
})();

File: crates/runtime/core/src/primitives/virtualizer.rs
//! Virtualizer — type-erased windowed/recycled list primitive.
//!
//! Authors don't call this directly. They use the generic
//! `flat_list<T>(...)` wrapper in `primitives::flat_list`, which
//! captures their typed data + closures and feeds the type-erased
//! callbacks Virtualizer needs.
//!
//! # Per-backend strategy
//!
//! - **Web**: a JS-side scroll handler (`backend-web/runtime/ts/virtualizer.ts`)
//!   owns the scroll listener + visible-range diff. It calls back
//!   into Rust only when items enter/leave the window. Per-item
//!   scopes are dropped on exit, so signals/effects for unmounted
//!   items are freed.
//! - **iOS**: `UICollectionView` with a flow layout that consults
//!   our `item_height` callback. Real cell recycling: `prepareForReuse`
//!   releases the item subtree, `cellForItemAt` builds the next one.
//! - **Android**: `RecyclerView` with a `ListAdapter` + `DiffUtil`.
//!   `onBindViewHolder` calls Rust to build the subtree;
//!   `onViewRecycled` releases it.
//!
//! All three backends see the same Rust contract.
//!
//! # Stable identity
//!
//! Every item carries a stable `u64` key (typically a hash of its
//! database id or content). When the data changes, the framework
//! diffs old keys against new keys to decide what to preserve.
//! Items whose key still exists keep their mounted subtree intact
//! — they may move in the layout, but their internal signals,
//! refs, and mounted state survive.
//!
//! # Size resolution
//!
//! Two modes per `ItemSize`:
//! - `Known`: author provides exact size per item before mount.
//!   Layout is deterministic. Cheapest.
//! - `Measured`: author provides an *estimate* per item; backend
//!   measures the actual rendered size on mount and stores it.
//!   Subsequent layout uses the measured value. If the item's
//!   rendered size changes later (its content updated), the
//!   backend's layout-observation primitive (ResizeObserver on web,
//!   `layoutSubviews` on iOS, `OnLayoutChangeListener` on Android)
//!   re-fires and refreshes the stored size.

use crate::{Bound, Element, Ref, RefFill};
use std::rc::Rc;

// The data/handle/Ops types of this primitive moved to `runtime-shared`
// (the walker-free half); this file keeps the Element/Bound builder
// surface (and its tests). The wildcard re-export preserves every old
// path.
pub use runtime_shared::primitives::virtualizer::*;

/// Construct a Virtualizer. Authors typically don't call this
/// directly — `flat_list<T>(...)` is the typed wrapper.
///
/// All callbacks are type-erased: the wrapper holds the actual
/// `T`-typed closures and bridges through these `usize`-only
/// callbacks. The framework's build path handles per-item scope
/// management — `render_item(idx)` runs inside a fresh `Scope`,
/// and that scope is dropped when the item is released.
#[cfg(feature = "prim-virtualizer")]
pub fn virtualizer(
    item_count: Box<dyn Fn() -> usize>,
    item_key: Box<dyn Fn(usize) -> ItemKey>,
    item_size: ItemSize,
    render_item: Rc<dyn Fn(usize) -> Element>,
) -> Bound<VirtualizerHandle> {
    // Closure-driven entry point: produce a `Derived<usize>` with
    // empty metadata (`is_opaque() == true`) so runtime backends
    // pick up the closure but generator backends report a clear
    // build-time error.
    let item_count = crate::derive::IntoDerived::<usize>::into_derived(item_count);
    Bound::new(Element::Virtualizer {
        item_count,
        item_key,
        item_size,
        render_item,
        row_template: None,
        row_index_signal_id: None,
        overscan: 1.0,
        layout: VirtualLayout::default(),
        style: None,
        ref_fill: None,
        accessibility: crate::accessibility::AccessibilityProps::default(),
    })
}

impl Bound<VirtualizerHandle> {
    /// Buffer factor outside the visible window. Default `1.0`
    /// (one viewport extent above and below). Higher = smoother
    /// fast-scroll, more memory.
    pub fn overscan(mut self, factor: f32) -> Self {
        if let Element::Virtualizer { overscan, .. } = &mut self.primitive {
            *overscan = factor;
        }
        self
    }

    /// Scroll axis. Default `Axis::Vertical`. `Axis::Horizontal`
    /// gives a horizontally-scrolling list/grid.
    pub fn axis(mut self, axis: Axis) -> Self {
        if let Element::Virtualizer { layout, .. } = &mut self.primitive {
            layout.axis = axis;
        }
        self
    }

    /// Cross-axis lane subdivision. `Lanes::Fixed(1)` (default) is a
    /// list; `Lanes::Fixed(N)` an N-lane uniform grid;
    /// `Lanes::AutoFit { min_cross }` a responsive grid.
    pub fn lanes(mut self, lanes: Lanes) -> Self {
        if let Element::Virtualizer { layout, .. } = &mut self.primitive {
            layout.lanes = lanes;
        }
        self
    }

    /// Gaps: `main` between successive grid-rows along the scroll
    /// axis, `cross` between lanes. For a list, only `main` (the
    /// inter-row gap) is meaningful.
    pub fn spacing(mut self, main: f32, cross: f32) -> Self {
        if let Element::Virtualizer { layout, .. } = &mut self.primitive {
            layout.main_spacing = main;
            layout.cross_spacing = cross;
        }
        self
    }

    /// Convenience: equal gap on both axes.
    pub fn gap(self, gap: f32) -> Self {
        self.spacing(gap, gap)
    }

    pub fn bind(mut self, r: Ref<VirtualizerHandle>) -> Self {
        if let Element::Virtualizer { ref_fill, .. } = &mut self.primitive {
            *ref_fill = Some(RefFill::Virtualizer(Box::new(move |h| r.fill(h))));
        }
        self
    }
}

#[cfg(all(test, feature = "prim-virtualizer"))]
mod tests {
    use super::*;

    fn empty_virtualizer() -> Bound<VirtualizerHandle> {
        virtualizer(
            Box::new(|| 0),
            Box::new(|i| i as u64),
            ItemSize::Known(Rc::new(|_| 40.0)),
            Rc::new(|_| crate::view(Vec::new()).primitive),
        )
    }

    fn layout_of(b: &Bound<VirtualizerHandle>) -> VirtualLayout {
        match &b.primitive {
            Element::Virtualizer { layout, .. } => *layout,
            _ => unreachable!("virtualizer() builds Element::Virtualizer"),
        }
    }

    #[test]
    fn default_layout_is_vertical_single_lane_list() {
        let l = layout_of(&empty_virtualizer());
        assert_eq!(l.axis, Axis::Vertical);
        assert_eq!(l.lanes, Lanes::Fixed(1));
        assert_eq!(l.main_spacing, 0.0);
        assert_eq!(l.cross_spacing, 0.0);
        assert!(!l.is_grid());
    }

    #[test]
    fn builders_set_axis_lanes_and_spacing() {
        let b = empty_virtualizer()
            .axis(Axis::Horizontal)
            .lanes(Lanes::Fixed(4))
            .spacing(6.0, 10.0);
        let l = layout_of(&b);
        assert_eq!(l.axis, Axis::Horizontal);
        assert_eq!(l.lanes, Lanes::Fixed(4));
        assert_eq!(l.main_spacing, 6.0);
        assert_eq!(l.cross_spacing, 10.0);
        assert!(l.is_grid());
    }

    #[test]
    fn gap_sets_both_axes_equally() {
        let l = layout_of(&empty_virtualizer().gap(12.0));
        assert_eq!(l.main_spacing, 12.0);
        assert_eq!(l.cross_spacing, 12.0);
    }

    #[test]
    fn fixed_lanes_ignore_container_extent() {
        // Fixed always returns its count (>=1), regardless of cross size.
        assert_eq!(Lanes::Fixed(3).resolve(1000.0, 8.0), 3);
        assert_eq!(Lanes::Fixed(3).resolve(0.0, 8.0), 3);
        // Zero clamps up to one — never a divide-by-zero grid.
        assert_eq!(Lanes::Fixed(0).resolve(1000.0, 0.0), 1);
    }

    #[test]
    fn autofit_resolves_gap_aware_lane_count() {
        // 5 lanes of 100 + 4 gaps of 10 = 540 <= 540: exactly 5.
        assert_eq!(
            Lanes::AutoFit { min_cross: 100.0 }.resolve(540.0, 10.0),
            5
        );
        // One pixel short of fitting a 5th lane → 4.
        assert_eq!(
            Lanes::AutoFit { min_cross: 100.0 }.resolve(539.0, 10.0),
            4
        );
        // No gaps: floor(cross / min).
        assert_eq!(
            Lanes::AutoFit { min_cross: 160.0 }.resolve(500.0, 0.0),
            3
        );
    }

    #[test]
    fn autofit_degrades_to_one_lane_on_unknown_or_narrow_container() {
        // Container narrower than one min lane → 1 (a list), never 0.
        assert_eq!(Lanes::AutoFit { min_cross: 200.0 }.resolve(150.0, 0.0), 1);
        // Zero/unknown container extent → 1.
        assert_eq!(Lanes::AutoFit { min_cross: 200.0 }.resolve(0.0, 0.0), 1);
        // Nonsense min → 1.
        assert_eq!(Lanes::AutoFit { min_cross: 0.0 }.resolve(500.0, 0.0), 1);
    }

    #[test]
    fn autofit_is_always_reported_as_grid() {
        // Even though AutoFit can resolve to one lane at runtime, the
        // author asked for grid behavior, so `is_grid()` is true.
        let b = empty_virtualizer().lanes(Lanes::AutoFit { min_cross: 120.0 });
        assert!(layout_of(&b).is_grid());
    }
}
